protocol/app/upgrades_test.go

```diff
@@ -0,0 +1,80 @@
+package app_test
+
+import (
+	"errors"
+	"testing"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	"github.com/dydxprotocol/v4-chain/protocol/app"
+	"github.com/dydxprotocol/v4-chain/protocol/app/upgrades"
+	testapp "github.com/dydxprotocol/v4-chain/protocol/testutil/app"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSetupUpgradeHandlers(t *testing.T) {
+	upgradeName := "test-upgrade"
+	createHandlerWasCalled := 0
+	handlerWasCalled := 0
+	upgradeHandlerFn := func(manager *module.Manager, configurator module.Configurator) types.UpgradeHandler {
+		createHandlerWasCalled++
+		return func(ctx sdk.Context, plan types.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+			handlerWasCalled++
+			return nil, errors.New("upgrade failed")
+		}
+	}
+
+	tests := map[string]struct {
+		upgrades      []upgrades.Upgrade
+		expectedPanic string
+	}{
+		"Successfully setup handler": {
```

Comment:
Suggestions for other test cases:
* what about multiple handlers that are successful?
* what about handlers that target the same height? (maybe this is testing upgrade keeper's behavior so it's not worth the test here)

Reply:
Added multiple handlers that are successful. I think testing upgrade keepers behavior makes sense as an integration test that does perform a full upgrade.